// chemicaltools/src/info/elementinfo.js
var origin = [
  {
    1: "氢",
    2: "H",
    3: 1.008,
    4: "Hydrogen",
    5: "Composed of the Greek elements hydro-and -gen meaning water-forming",
  },
  {
    1: "氦",
    3: 4.003,
    4: "Helium",
    5: "The Greek helios, sun",
  },
  {
    1: "锂",
    3: 6.941,
    4: "Lithium",
    5: "The Greek lithos, stone",
  },
  {
    1: "铍",
    3: 9.012,
    4: "Beryllium",
    5: "Beryl, a mineral",
  },
  {
    1: "硼",
    2: "B",
    3: 10.81,
    4: "Boron",
    5: "Borax, a mineral",
  },
  {
    1: "碳",
    2: "C",
    3: 12.01,
    4: "Carbon",
    5: "The Latin carbo, coal",
  },
  {
    1: "氮",
    2: "N",
    3: 14.01,
    4: "Nitrogen",
    5: "The Greek nitron and -gen meaning niter-forming",
  },
  {
    1: "氧",
    2: "O",
    3: 16,
    4: "Oxygen",
    5: "From the Greek oxy-, both sharp and acid, and -gen, meaning acid-forming",
  },
  {
    1: "氟",
    2: "F",
    3: 19,
    4: "Fluorine",
    5: "The Latin fluere, to flow",
  },
  {
    1: "氖",
    3: 20.18,
    4: "Neon",
    5: "The Greek neos, meaning new",
  },
  {
    1: "钠",
    2: "Na",
    3: 22.99,
    4: "Sodium",
    5: "The English word soda (natrium in Latin)",
  },
  {
    1: "镁",
    2: "Mg",
    3: 24.31,
    4: "Magnesium",
    5: "Magnesia, a district of Eastern Thessalyin Greece",
  },
  {
    1: "铝",
    3: 26.98,
    4: "Aluminium",
    5: "From alumina, a compound (originally aluminum)",
  },
  {
    1: "硅",
    3: 28.09,
    4: "Silicon",
    5: "From the Latin silex, flint (originally silicium)",
  },
  {
    1: "磷",
    2: "P",
    3: 30.98,
    4: "Phosphorus",
    5: "The Greek phoosphoros, carrying light",
  },
  {
    1: "硫",
    2: "S",
    3: 32.07,
    4: "Sulfur",
    5: "The Latin sulphur, fire and brimstone",
  },
  {
    1: "氯",
    2: "Cl",
    3: 35.45,
    4: "Chlorine",
    5: "The Greek chloros, greenish yellow",
  },
  {
    1: "氩",
    3: 39.95,
    4: "Argon",
    5: "The Greek argos, idle",
  },
  {
    1: "钾",
    2: "K",
    3: 39.1,
    4: "Potassium",
    5: "New Latin potassa, potash (kalium in Latin)",
  },
  {
    1: "钙",
    3: 40.08,
    4: "Calcium",
    5: "The Latin calx, lime",
  },
  {
    1: "钪",
    3: 44.96,
    4: "Scandium",
    5: "Scandia, the Latin name for Scandinavia",
  },
  {
    1: "钛",
    3: 47.87,
    4: "Titanium",
    5: "Titans, the sons of the Earth goddess of Greek mythology",
  },
  {
    1: "钒",
    2: "V",
    3: 50.94,
    4: "Vanadium",
    5: "Vanadis, an Old Norse name for the Scandinavian goddess Freyja",
  },
  {
    1: "铬",
    2: "Cr",
    3: 52,
    4: "Chromium",
    5: "The Greek chroma, color",
  },
  {
    1: "锰",
    2: "Mn",
    3: 54.94,
    4: "Manganese",
    5: "Corrupted from magnesia negra, see Magnesium",
  },
  {
    1: "铁",
    2: "Fe",
    3: 55.85,
    4: "Iron",
    5: "English word (ferrum in Latin)",
  },
  {
    1: "钴",
    3: 58.93,
    4: "Cobalt",
    5: "The German word Kobold, goblin",
  },
  {
    1: "镍",
    3: 58.69,
    4: "Nickel",
    5: "From Swedish kopparnickel, containing the German word Nickel, goblin",
  },
  {
    1: "铜",
    2: "Cu",
    3: 63.55,
    4: "Copper",
    5: "English word (Latin cuprum)",
  },
  {
    1: "锌",
    2: "Zn",
    3: 65.38,
    4: "Zinc",
    5: "The German Zink",
  },
  {
    1: "镓",
    3: 69.72,
    4: "Gallium",
    5: "Gallia, the Latin name for France",
  },
  {
    1: "锗",
    3: 72.64,
    4: "Germanium",
    5: "Germania, the Latin name for Germany",
  },
  {
    1: "砷",
    2: "As",
    3: 74.92,
    4: "Arsenic",
    5: "English word (Latin arsenicum)",
  },
  {
    1: "硒",
    3: 78.97,
    4: "Selenium",
    5: "The Greek selene, moon",
  },
  {
    1: "溴",
    3: 79.9,
    4: "Bromine",
    5: "The Greek bromos, stench",
  },
  {
    1: "氪",
    3: 83.8,
    4: "Krypton",
    5: "The Greek kryptos, hidden",
  },
  {
    1: "铷",
    2: "Rb",
    3: 85.47,
    4: "Rubidium",
    5: "The Latin rubidus, deep red",
  },
  {
    1: "锶",
    2: "Sr",
    3: 87.62,
    4: "Strontium",
    5: "Strontian, a small town in Scotland",
  },
  {
    1: "钇",
    2: "Y",
    3: 88.91,
    4: "Yttrium",
    5: "Ytterby, Sweden",
  },
  {
    1: "锆",
    2: "Zr",
    3: 91.22,
    4: "Zirconium",
    5: "Persian Zargun, gold-colored; German Zirkoon, jargoon",
  },
  {
    1: "铌",
    2: "Nb",
    3: 92.91,
    4: "Niobium",
    5: "Niobe, daughter of king Tantalus from Greek mythology",
  },
  {
    1: "钼",
    3: 95.95,
    4: "Molybdenum",
    5: "The Greek molybdos meaning lead",
  },
  {
    1: "锝",
    2: "Tc",
    3: 98.91,
    4: "Technetium",
    5: "The Greek tekhnètos meaning artificial",
  },
  {
    1: "钌",
    3: 101.1,
    4: "Ruthenium",
    5: "Ruthenia, the New Latin name for Russia",
  },
  {
    1: "铑",
    3: 102.9,
    4: "Rhodium",
    5: "The Greek rhodos, meaning rose coloured",
  },
  {
    1: "钯",
    2: "Pd",
    3: 106.4,
    4: "Palladium",
    5: "The then recently discovered asteroid Pallas, considered a planet at the time",
  },
  {
    1: "银",
    2: "Ag",
    3: 107.9,
    4: "Silver",
    5: "English word (argentum in Latin)",
  },
  {
    1: "镉",
    2: "Cd",
    3: 112.4,
    4: "Cadmium",
    5: "The New Latin cadmia, from King Kadmos",
  },
  {
    1: "铟",
    3: 114.8,
    4: "Indium",
    5: "Indigo",
  },
  {
    1: "锡",
    2: "Sn",
    3: 118.7,
    4: "Tin",
    5: "English word (stannum in Latin)",
  },
  {
    1: "锑",
    2: "Sb",
    3: 121.8,
    4: "Antimony",
    5: "Composed from the Greek anti, against, and monos, alone (stibium in Latin)",
  },
  {
    1: "碲",
    3: 127.6,
    4: "Tellurium",
    5: "Latin tellus, earth",
  },
  {
    1: "碘",
    2: "I",
    3: 126.9,
    4: "Iodine",
    5: "French iode (after the Greek ioeides, violet)",
  },
  {
    1: "氙",
    3: 131.3,
    4: "Xenon",
    5: "The Greek xenos, strange",
  },
  {
    1: "铯",
    2: "Cs",
    3: 132.9,
    4: "Caesium",
    5: "The Latin caesius, sky blue",
  },
  {
    1: "钡",
    3: 137.3,
    4: "Barium",
    5: "The Greek barys, heavy",
  },
  {
    1: "镧",
    3: 138.9,
    4: "Lanthanum",
    5: "The Greek lanthanein, to lie hidden",
  },
  {
    1: "铈",
    3: 140.1,
    4: "Cerium",
    5: "The then recently discovered asteroid Ceres, considered a planet at the time",
  },
  {
    1: "镨",
    3: 140.9,
    4: "Praseodymium",
    5: "The Greek praseios didymos meaning green twin",
  },
  {
    1: "钕",
    2: "Nd",
    3: 144.2,
    4: "Neodymium",
    5: "The Greek neos didymos meaning new twin",
  },
  {
    1: "钷",
    2: "Pm",
    3: 144.9,
    4: "Promethium",
    5: "Prometheus of Greek mythology who stole fire from the Gods and gave it to humans",
  },
  {
    1: "钐",
    2: "Sm",
    3: 150.4,
    4: "Samarium",
    5: "Samarskite, the name of the mineral from which it was first isolated",
  },
  {
    1: "铕",
    3: 152,
    4: "Europium",
    5: "Europe",
  },
  {
    1: "钆",
    2: "Gd",
    3: 157.3,
    4: "Gadolinium",
    5: "Johan Gadolin, chemist, physicist and mineralogist",
  },
  {
    1: "铽",
    2: "Tb",
    3: 158.9,
    4: "Terbium",
    5: "Ytterby, Sweden",
  },
  {
    1: "镝",
    3: 162.5,
    4: "Dysprosium",
    5: "The Greek dysprositos, hard to get",
  },
  {
    1: "钬",
    3: 164.9,
    4: "Holmium",
    5: "Holmia, the New Latin name for Stockholm",
  },
  {
    1: "铒",
    3: 167.3,
    4: "Erbium",
    5: "Ytterby, Sweden",
  },
  {
    1: "铥",
    2: "Tm",
    3: 168.9,
    4: "Thulium",
    5: "Thule, the ancient name for Scandinavia",
  },
  {
    1: "镱",
    2: "Yb",
    3: 173.1,
    4: "Ytterbium",
    5: "Ytterby, Sweden",
  },
  {
    1: "镥",
    3: 175,
    4: "Lutetium",
    5: "Lutetia, the Latin name for Paris",
  },
  {
    1: "铪",
    2: "Hf",
    3: 178.5,
    4: "Hafnium",
    5: "Hafnia, the New Latin name for Copenhagen",
  },
  {
    1: "钽",
    3: 181,
    4: "Tantalum",
    5: "King Tantalus, father of Niobe from Greek mythology",
  },
  {
    1: "钨",
    2: "W",
    3: 183.8,
    4: "Tungsten",
    5: "The Swedish tung sten, heavy stone (W is wolfram, the old name of the tungsten mineral wolframite)",
  },
  {
    1: "铼",
    2: "Re",
    3: 186.2,
    4: "Rhenium",
    5: "Rhenus, the Latin name for the river Rhine",
  },
  {
    1: "锇",
    3: 190.2,
    4: "Osmium",
    5: "The Greek osmè, meaning smell",
  },
  {
    1: "铱",
    3: 192.2,
    4: "Iridium",
    5: "Iris, the Greek goddess of the rainbow",
  },
  {
    1: "铂",
    2: "Pt",
    3: 195.1,
    4: "Platinum",
    5: "The Spanish platina, meaning little silver",
  },
  {
    1: "金",
    2: "Au",
    3: 197,
    4: "Gold",
    5: "English word (aurum in Latin)",
  },
  {
    1: "汞",
    2: "Hg",
    3: 200.6,
    4: "Mercury",
    5: "The New Latin name mercurius, named after the Roman god (Hg from former name hydrargyrum, from Greek hydr-, water, and argyros, silver)",
  },
  {
    1: "铊",
    2: "Tl",
    3: 204.4,
    4: "Thallium",
    5: "The Greek thallos, green twig",
  },
  {
    1: "铅",
    2: "Pb",
    3: 207.2,
    4: "Lead",
    5: "English word (plumbum in Latin)",
  },
  {
    1: "铋",
    3: 209,
    4: "Bismuth",
    5: "German word, now obsolete",
  },
  {
    1: "钋",
    3: 209,
    4: "Polonium",
    5: "Polonia, the New Latin name for Poland",
  },
  {
    1: "砹",
    2: "At",
    3: 210,
    4: "Astatine",
    5: "The Greek astatos, unstable",
  },
  {
    1: "氡",
    2: "Rn",
    3: 222,
    4: "Radon",
    5: "From radium, as it was first detected as an emission from radium during radioactive decay",
  },
  {
    1: "钫",
    3: 223,
    4: "Francium",
    5: "Francia, the New Latin name for France",
  },
  {
    1: "镭",
    2: "Re",
    3: 226,
    4: "Radium",
    5: "The Latin radius, ray",
  },
  {
    1: "锕",
    3: 227,
    4: "Actinium",
    5: "The Greek aktis, ray",
  },
  {
    1: "钍",
    3: 232,
    4: "Thorium",
    5: "Thor, the Scandinavian god of thunder",
  },
  {
    1: "镤",
    2: "Pa",
    3: 231,
    4: "Protactinium",
    5: "The Greek protos, first, and actinium, which is produced through the radioactive decay of protactinium",
  },
  {
    1: "铀",
    2: "U",
    3: 238,
    4: "Uranium",
    5: "Uranus, the seventh planet in the Solar System",
  },
  {
    1: "镎",
    2: "Np",
    3: 237,
    4: "Neptunium",
    5: "Neptune, the eighth planet in the Solar System",
  },
  {
    1: "钚",
    2: "Pu",
    3: 239.1,
    4: "Plutonium",
    5: "Pluto, a dwarf planet in the Solar System (considered the ninth planet at the time)",
  },
  {
    1: "镅",
    3: 243.1,
    4: "Americium",
    5: "The Americas, as the element was first synthesized on the continent, by analogy with europium",
  },
  {
    1: "锔",
    2: "Cm",
    3: 247.1,
    4: "Curium",
    5: "Pierre Curie, a physicist, and Marie Curie, a physicist and chemist, named after great scientists by analogy with gadolinium",
  },
  {
    1: "锫",
    2: "Bk",
    3: 247.1,
    4: "Berkelium",
    5: "Berkeley, California, where the element was first synthesized, by analogy with terbium",
  },
  {
    1: "锎",
    2: "Cf",
    3: 251.1,
    4: "Californium",
    5: "California, where the element was first synthesized",
  },
  {
    1: "锿",
    2: "Es",
    3: 252.1,
    4: "Einsteinium",
    5: "Albert Einstein, physicist",
  },
  {
    1: "镄",
    2: "Fm",
    3: 257.1,
    4: "Fermium",
    5: "Enrico Fermi, physicist",
  },
  {
    1: "钔",
    2: "Md",
    3: 258.1,
    4: "Mendelevium",
    5: "Dmitri Mendeleev, chemist and inventor",
  },
  {
    1: "锘",
    3: 259.1,
    4: "Nobelium",
    5: "Alfred Nobel, chemist, engineer, innovator, and armaments manufacturer",
  },
  {
    1: "铹",
    2: "Lr",
    3: 262.1,
    4: "Lawrencium",
    5: "Ernest O. Lawrence, physicist",
  },
  {
    1: "𬬻",
    2: "Rf",
    3: 261.1,
    4: "Rutherfordium",
    5: "Ernest Rutherford, chemist and physicist",
    6: "lu",
  },
  {
    1: "𬭊",
    2: "Db",
    3: 262.1,
    4: "Dubnium",
    5: "Dubna, Russia",
    6: "du",
  },
  {
    1: "𬭳",
    2: "Sg",
    3: 266.1,
    4: "Seaborgium",
    5: "Glenn T. Seaborg, scientist",
    6: "xi",
  },
  {
    1: "𬭛",
    2: "Bh",
    3: 264.1,
    4: "Bohrium",
    5: "Niels Bohr, physicist",
    6: "bo",
  },
  {
    1: "𬭶",
    2: "Hs",
    3: 277,
    4: "Hassium",
    5: "Hesse, Germany, where the element was first synthesized",
    6: "hei",
  },
  {
    1: "鿏",
    2: "Mt",
    3: 268,
    4: "Meitnerium",
    5: "Lise Meitner, physicist",
    6: "mai",
  },
  {
    1: "𫟼",
    2: "Ds",
    3: 271,
    4: "Darmstadtium",
    5: "Darmstadt, Germany, where the element was first synthesized",
    6: "da",
  },
  {
    1: "𬬭",
    2: "Rg",
    3: 272,
    4: "Roentgenium",
    5: "Wilhelm Conrad Röntgen, physicist",
    6: "lun",
  },
  {
    1: "鿔",
    2: "Cn",
    3: 285,
    4: "Copernicium",
    5: "Nicolaus Copernicus, astronomer",
    6: "ge",
  },
  {
    1: "鿭",
    2: "Nh",
    3: 284,
    4: "Nihonium",
    5: "Japanese word meaning Japan",
  },
  {
    1: "𫓧",
    3: 289,
    4: "Flerovium",
    5: "Georgy Flyorov, physicist",
    6: "fu",
  },
  {
    1: "镆",
    2: "Mc",
    3: 288,
    4: "Moscovium",
    5: "In recognition of the Moscow region and honors the ancient Russian land that is the home of the Joint Institute for Nuclear Research",
  },
  {
    1: "𫟷",
    2: "Lv",
    3: 292,
    4: "Livermorium",
    5: "Lawrence Livermore National Laboratory(in Livermore, California) which collaborated with JINR on its synthesis",
    6: "li",
  },
  {
    1: "鿬",
    2: "Ts",
    3: 295,
    4: "Tennessine",
    5: "In recognition of the contribution of the Tennessee region",
    6: "tian",
  },
  {
    1: "鿫",
    3: 293,
    4: "Oganesson",
    5: "Honoring a scientist and recognizes Professor Yuri Oganessian",
    6: "ao",
  }
];

var pinyin = require("tiny-pinyin");
var getnew = function(ori) {
  var n = [];
  for (var i in ori) {
    var info = ori[i];
    var newinfo = {};
    newinfo["number"] = parseInt(i) + 1;
    newinfo["name"] = info[1];
    if (info[2]) {
      newinfo["symbol"] = info[2];
    } else {
      // default is the first two letters in the iupac
      newinfo["symbol"] = info[4][0] + info[4][1];
    }
    newinfo["mass"] = info[3];
    newinfo["iupac"] = info[4];
    newinfo["origin"] = info[5];
    if (info[6]) {
      newinfo["pinyin"] = info[6];
    } else {
      try {
        newinfo["pinyin"] = pinyin.convertToPinyin(info[1]).toLowerCase();
      } catch (e) {
        newinfo["pinyin"] = "";
      }
    }
    newinfo["url"] = "https://cdn.jsdelivr.net/gh/njzjz/chemicaltools-android@ee2112766bbf2596a7623edc64a2c1c4ec8ff3be/app/src/main/res/drawable/element_" + newinfo["number"].toString() + ".png";
    n[i] = newinfo;
  }
  return n;
};
var elementinfo = getnew(origin);
var search = function(kind, x) {
  for (var i in elementinfo) {
    var info = elementinfo[i];
    if (info[kind] && x && String(info[kind]).toLowerCase() == x.toLowerCase()) {
      return info;
    }
  }
  return null;
};
exports.elementinfo = elementinfo;
exports.search = search;
